src/System.Private.Windows.Core/src/System/Windows/Forms/BinaryFormat/Support/SerializationExtensions.cs

```diff
@@ -42,4 +42,51 @@ public static Span<T> GetArrayData<T>(this Array array)
 
         return data;
     }
+
+    /// <summary>
+    ///  Sets a value in an array by its flattened index.
+    /// </summary>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static void SetArrayValueByFlattenedIndex(this Array array, object? value, int flattenedIndex)
+    {
+        // Fast path
+        int ranks = array.Rank;
+        if (ranks == 1)
+        {
+            array.SetValue(value, flattenedIndex);
+            return;
+        }
+
+        SetArrayValueByFlattenedIndex(array, flattenedIndex, value);
+
+        static void SetArrayValueByFlattenedIndex(Array array, int flattenedIndex, object? value)
+        {
+            int ranks = array.Rank;
+
+            if (ranks == 2)
+            {
+                int length = array.GetLength(1);
+                int index1 = flattenedIndex / length;
+                int index2 = flattenedIndex - (index1 * length);
```

Comment:
Isn't this just

```suggestion
                (int index1, int index2) = int.DivRem(flattenedIndex, length);
```

I mean, this is fine... but (IIUC) Div and DivRem are the same instruction, so might as well use the one instruction to get both parts and save on an extra multiplication and subtraction.